AdvancedC#/FileSystemVisitorPL/FileSystemVisitor.cs

```diff
@@ -28,15 +54,31 @@ public FileSystemVisitor(string rootPath, Func<string, bool> customFilter)
 
         public IEnumerable<string> GetFilesAndFolders()
         {
-            return TraverseDirectory(rootFolder);
+            OnSearchStarted();
+
+            foreach (var item in TraverseDirectory(rootFolder))
+            {
+                yield return item;
+            }
+
+            OnSearchFinished();
         }
 
         private IEnumerable<string> TraverseDirectory(string directory)
         {
             if (searchFilter(directory))
             {
+                var args = new FileSystemEventArgs(directory);
+                OnFilteredDirectoryFound(args);
+                if (args.ExcludeItem) yield break;
+                if (args.AbortSearch) OnSearchFinished();
```

Comment:
1. Looks like you can put this items into separate function since it repeats from 95 to 98 rows
2. Don't forget to wrap conditions with brackets, it simplifies readability